### PR TITLE
Validate writer arguments

### PR DESCRIPTION
**Pull request by karan6181**
## Description of changes:
- Validate Writer arguments and raise a ValueError exception if argument(s) is/are invalid

## Issue #, if available:
STR-135

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [ ] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#submitting-a-contribution))
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

**Review by dakinggg — APPROVED** [review]
lgtm

**Review by Skylion007 — APPROVED** [review]
Amazing!

**Comment by xiaohanzhan-db** [comment]
LGTM. Since we check keys now, shall we also check type of values as well? something like
        
allowed_extra_args = {'progress_bar': bool, 'max_workers': int}
invalid_args = [arg for arg, value in kwargs.items() if not (arg in allowed_extra_args and allowed_extra_args[arg] == type(value))]

File: streaming/base/format/base/writer.py [hunk 1]
```diff
@@ -94,6 +94,13 @@ def __init__(self,
                 raise ValueError(f'`size_limit` must be greater than zero, instead, ' +
                                  f'found as {size_limit_value}.')
 
+        # Validate keyword arguments
+        invalid_kwargs = [
+            args for args in kwargs.keys() if args not in ('progress_bar', 'max_workers')
```
Comment (knighton) [inline streaming/base/format/base/writer.py:99]: `args` -> `arg`

why kwargs.keys()? isn't .keys() pointless here

File: scripts/samples/bench_and_plot.py [hunk 2]
```diff
@@ -339,13 +339,12 @@ def bench(args: Namespace, bench_name: str, desc: str, generate: Callable,
     plt.ylabel('Probability')
     plt.grid(which='major', ls='-', c='#ddd', lw=0.5)
     plt.grid(which='minor', ls=':', c='#ddd', lw=0.5)
+    formatter = ScalarFormatter()
```
Comment (karan6181) [inline scripts/samples/bench_and_plot.py:342]: Pyright was complaining `error: Cannot access member "set_scientific" for type "Formatter"`. Hence, this change.

File: streaming/base/format/base/writer.py (hunk 1 above)
Comment (karan6181) [reply streaming/base/format/base/writer.py]: updated. Kept `keys()` just to be more specific. No harm either ways.

**Comment by karan6181** [comment]
> LGTM. Since we check keys now, shall we also check type of values as well? something like
> 
> allowed_extra_args = {'progress_bar': bool, 'max_workers': int} invalid_args = [arg for arg, value in kwargs.items() if not (arg in allowed_extra_args and allowed_extra_args[arg] == type(value))]

Great point. Would prefer doing it in a separate PR. Also, would like to understand on what level of check type do we need in our codebase!

**Pull request merged**